File: src/codeq/service.py
from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .contracts import attach_schema
from .util import git_root
from .workspace import Workspace


@dataclass
class _WorkspaceEntry:
    workspace: Workspace
    last_used: float
    active: int = 0
    semantic_find_gate: threading.Lock = field(default_factory=threading.Lock)


class CodeqService:
    def __init__(self, max_workspaces: int = 4) -> None:
        self._workspaces: dict[Path, _WorkspaceEntry] = {}
        self._lock = threading.RLock()
        self._last_activity = time.monotonic()
        self._max_workspaces = max(1, max_workspaces)

    def close(self) -> None:
        with self._lock:
            workspaces = [entry.workspace for entry in self._workspaces.values()]
            self._workspaces.clear()
        for workspace in workspaces:
            workspace.close()

    def workspace_count(self) -> int:
        with self._lock:
            return len(self._workspaces)

    def idle_seconds(self) -> float:
        with self._lock:
            return max(0.0, time.monotonic() - self._last_activity)

    def evict_idle(self, max_idle_seconds: float, *, lsp_idle_seconds: float | None = None) -> list[str]:
        """Close inactive workspaces without exposing lifecycle controls to the CLI."""
        now = time.monotonic()
        warm_limit = max_idle_seconds if lsp_idle_seconds is None else lsp_idle_seconds
        evicted: list[Workspace] = []
        roots: list[str] = []
        with self._lock:
            for root, entry in list(self._workspaces.items()):
                idle_limit = warm_limit if entry.workspace.has_active_lsp() else max_idle_seconds
                if entry.active != 0 or now - entry.last_used < idle_limit:
                    continue
                self._workspaces.pop(root, None)
                evicted.append(entry.workspace)
                roots.append(str(root))
        for workspace in evicted:
            workspace.close()
        return roots

    def _acquire_workspace(self, root: str, timeout: float) -> tuple[Path, _WorkspaceEntry]:
        resolved = git_root(root)
        now = time.monotonic()
        victim: Workspace | None = None
        with self._lock:
            entry = self._workspaces.get(resolved)
            if entry is None:
                if len(self._workspaces) >= self._max_workspaces:
                    idle = [
                        (candidate_root, candidate)
                        for candidate_root, candidate in self._workspaces.items()
                        if candidate.active == 0
                    ]
                    if idle:
                        victim_root, victim_entry = min(idle, key=lambda item: item[1].last_used)
                        self._workspaces.pop(victim_root, None)
                        victim = victim_entry.workspace
                entry = _WorkspaceEntry(Workspace(resolved, timeout=timeout), last_used=now)
                self._workspaces[resolved] = entry
            else:
                entry.workspace.timeout = timeout
            entry.active += 1
            entry.last_used = now
            self._last_activity = now
        if victim is not None:
            victim.close()
        return resolved, entry

    def _release_workspace(self, entry: _WorkspaceEntry) -> None:
        with self._lock:
            entry.active = max(0, entry.active - 1)
            entry.last_used = time.monotonic()
            self._last_activity = entry.last_used

    def handle(self, request: dict[str, Any]) -> dict[str, Any]:
        started = time.perf_counter()
        command = request.get("command")
        if command == "_status":
            with self._lock:
                self._last_activity = time.monotonic()
                roots = sorted(str(root) for root in self._workspaces)
            return attach_schema({"status": "ok", "workspaces": len(roots), "roots": roots})

        root = str(request.get("root") or ".")
        timeout = float(request.get("timeout") or 15.0)
        limit = int(request.get("limit") or 20)
        _, entry = self._acquire_workspace(root, timeout)
        ws = entry.workspace
        gate_acquired = False
        queue_ms = 0.0
        try:
            if command == "find" and not bool(request.get("text")):
                queued_at = time.perf_counter()
                gate_acquired = entry.semantic_find_gate.acquire(timeout=timeout)
                queue_ms = (time.perf_counter() - queued_at) * 1000
                if not gate_acquired:
                    raise TimeoutError(
                        f"semantic find timed out after {timeout:g}s waiting for another query"
                    )
            execution_started = time.perf_counter()
            before = ws.session_stats()
            metrics_before = ws.metrics_snapshot()
            data: dict[str, Any]
            if command == "find":
                data = ws.find(
                    str(request.get("query") or ""),
                    limit=limit,
                    kind=str(request.get("kind") or "") or None,
                    text=bool(request.get("text")),
                    paths=tuple(str(value) for value in (request.get("paths") or [])),
                    globs=tuple(str(value) for value in (request.get("globs") or [])),
                    exclude_tests=bool(request.get("exclude_tests")),
                )
            elif command == "context":
                data = ws.context(
                    str(request.get("target") or ""),
                    limit=limit,
                    outline_depth=max(0, int(request.get("outline_depth") or 1)),
                    outline_kind=str(request.get("outline_kind") or "") or None,
                    container=str(request.get("container") or "") or None,
                    include_topology=bool(request.get("include_topology")),
                    lexical_references=bool(request.get("lexical_references")),
                    lexical_query=str(request.get("lexical_query") or "") or None,
                    lexical_paths=tuple(str(value) for value in (request.get("lexical_paths") or [])),
                    lexical_globs=tuple(str(value) for value in (request.get("lexical_globs") or [])),
                    lexical_exclude_tests=bool(request.get("lexical_exclude_tests")),
                    semantic_paths=tuple(str(value) for value in (request.get("semantic_paths") or [])),
                )
            elif command == "trace":
                raw_depth = request.get("depth")
                depth = 3 if raw_depth is None else int(raw_depth)
                if depth < 0:
                    raise ValueError("trace depth must be >= 0")
                data = ws.trace(
                    str(request.get("target") or ""),
                    direction=str(request.get("direction") or "in"),
                    depth=depth,
                    limit=max(1, int(request.get("node_limit") or 100)),
                )
            elif command == "review":
                data = ws.review(
                    str(request.get("base") or "HEAD~1"),
                    limit=limit,
                    merge_base=bool(request.get("merge_base")),
                )
            else:
                raise ValueError(f"unknown command: {command}")
            phase_ms = data.pop("_phase_ms", None)
            after = ws.session_stats()
            metrics_after = ws.metrics_snapshot()

            def delta(name: str) -> int:
                return max(0, int(metrics_after.get(name, 0)) - int(metrics_before.get(name, 0)))

            finished = time.perf_counter()
            meta: dict[str, Any] = {
                "root": str(ws.root),
                "duration_ms": round((finished - started) * 1000, 1),
                "queue_ms": round(queue_ms, 1),
                "execution_ms": round((finished - execution_started) * 1000, 1),
                "lsp_sessions_before": before,
                "lsp_sessions": after,
                "lsp_started": delta("sessions_started") > 0,
                "lsp_request_count": delta("lsp_request_count"),
                "prewarm_files": delta("prewarm_files"),
                "prewarm_probes": delta("prewarm_probes"),
                "prewarm_early_stops": delta("prewarm_early_stops"),
                "cache": {
                    "document_symbols_hit": delta("document_symbols_hit"),
                    "document_symbols_miss": delta("document_symbols_miss"),
                    "document_symbols_waited": delta("document_symbols_waited"),
                    "document_symbols_evicted": delta("document_symbols_evicted"),
                    "document_symbol_entries": int(metrics_after.get("document_symbol_cache_entries", 0)),
                },
            }
            data["_meta"] = meta
            if isinstance(phase_ms, dict):
                meta["phase_ms"] = {
                    str(name): round(max(0.0, float(value)), 1)
                    for name, value in phase_ms.items()
                }
            if data.get("mode") == "text":
                meta["text"] = {
                    "matching_file_count": int(data.get("matching_file_count", 0)),
                    "tracked_matching_lines": int(data.get("tracked_line_count", 0)),
                    "untracked_matching_lines": int(data.get("untracked_line_count", 0)),
                }
            elif isinstance(data.get("lexical_references"), dict):
                lexical = data["lexical_references"]
                meta["text"] = {
                    "matching_file_count": int(lexical.get("matching_file_count", 0)),
                    "tracked_matching_lines": int(lexical.get("tracked_line_count", 0)),
                    "untracked_matching_lines": int(lexical.get("untracked_line_count", 0)),
                }
            return attach_schema(data)
        finally:
            if gate_acquired:
                entry.semantic_find_gate.release()
            self._release_workspace(entry)

File: src/service.rs
use std::collections::HashMap;
use std::collections::hash_map::DefaultHasher;
use std::fs;
use std::hash::{Hash, Hasher};
use std::path::Path;
use std::path::PathBuf;
use std::sync::{Arc, Condvar, Mutex, MutexGuard};
use std::time::Duration;
use std::time::Instant;

use serde_json::Value;

use crate::boundary;
use crate::cli::{Cli, Command, FindMode};
use crate::concept;
use crate::contracts::SCHEMA_VERSION;
use crate::gitreview;
use crate::semantic;
use crate::textsearch;
use crate::workspace::Workspace;

pub struct DaemonService {
    state: Mutex<DaemonState>,
    workspace_available: Condvar,
    max_workspaces: usize,
    scratch_root: PathBuf,
}

struct DaemonState {
    workspaces: HashMap<PathBuf, WorkspaceState>,
    last_activity: Instant,
}

struct WorkspaceState {
    active: usize,
    last_used: Instant,
    runtime: Arc<Workspace>,
}

struct WorkspaceLease<'a> {
    service: &'a DaemonService,
    root: PathBuf,
    runtime: Arc<Workspace>,
}

impl DaemonService {
    pub fn new(max_workspaces: usize, scratch_root: PathBuf) -> Self {
        Self {
            state: Mutex::new(DaemonState {
                workspaces: HashMap::new(),
                last_activity: Instant::now(),
            }),
            workspace_available: Condvar::new(),
            max_workspaces: max_workspaces.max(1),
            scratch_root,
        }
    }

    pub fn query(&self, cli: &Cli) -> Result<Value, String> {
        let root = cli.root.clone();
        let lease = self.acquire(&root, query_timeout(cli.timeout))?;
        Ok(execute_with_workspace(cli, &root, "daemon", &lease.runtime))
    }

    pub fn status(&self) -> Value {
        let mut state = self.lock_state();
        state.last_activity = Instant::now();
        let mut roots: Vec<_> = state
            .workspaces
            .keys()
            .map(|root| root.to_string_lossy().into_owned())
            .collect();
        roots.sort();
        serde_json::json!({
            "status": "ok",
            "workspaces": roots.len(),
            "roots": roots,
            "schema_version": SCHEMA_VERSION,
        })
    }

    pub fn evict_idle(&self, workspace_idle: Duration, lsp_idle: Duration) -> Vec<PathBuf> {
        let mut state = self.lock_state();
        let now = Instant::now();
        let evicted_roots: Vec<_> = state
            .workspaces
            .iter()
            .filter(|(_, workspace)| {
                let max_idle = if workspace.runtime.has_live_sessions() {
                    lsp_idle
                } else {
                    workspace_idle
                };
                workspace.active == 0 && now.duration_since(workspace.last_used) >= max_idle
            })
            .map(|(root, _)| root.clone())
            .collect();
        let evicted: Vec<_> = evicted_roots
            .iter()
            .filter_map(|root| state.workspaces.remove(root))
            .collect();
        if !evicted_roots.is_empty() {
            self.workspace_available.notify_all();
        }
        drop(state);
        drop(evicted);
        evicted_roots
    }

    pub fn workspace_count(&self) -> usize {
        self.lock_state().workspaces.len()
    }

    pub fn idle_duration(&self) -> Duration {
        Instant::now().duration_since(self.lock_state().last_activity)
    }

    fn acquire<'a>(&'a self, root: &Path, timeout: Duration) -> Result<WorkspaceLease<'a>, String> {
        let deadline = Instant::now() + timeout;
        let root = root.to_owned();
        let mut state = self.lock_state();
        loop {
            let now = Instant::now();
            if let Some(workspace) = state.workspaces.get_mut(&root) {
                workspace.active += 1;
                workspace.last_used = now;
                let runtime = Arc::clone(&workspace.runtime);
                state.last_activity = now;
                return Ok(WorkspaceLease {
                    service: self,
                    root,
                    runtime,
                });
            }
            if state.workspaces.len() >= self.max_workspaces {
                let victim = state
                    .workspaces
                    .iter()
                    .filter(|(_, workspace)| workspace.active == 0)
                    .min_by_key(|(_, workspace)| workspace.last_used)
                    .map(|(candidate, _)| candidate.clone());
                if let Some(victim) = victim {
                    let evicted = state.workspaces.remove(&victim);
                    drop(state);
                    drop(evicted);
                    state = self.lock_state();
                    continue;
                } else {
                    let remaining = deadline.saturating_duration_since(now);
                    if remaining.is_zero() {
                        return Err(format!(
                            "workspace capacity timed out after {}s",
                            timeout.as_secs_f64()
                        ));
                    }
                    let waited = self
                        .workspace_available
                        .wait_timeout(state, remaining)
                        .unwrap_or_else(|poisoned| poisoned.into_inner());
                    state = waited.0;
                    if waited.1.timed_out() {
                        return Err(format!(
                            "workspace capacity timed out after {}s",
                            timeout.as_secs_f64()
                        ));
                    }
                    continue;
                }
            }
            state.workspaces.insert(
                root.clone(),
                WorkspaceState {
                    active: 1,
                    last_used: now,
                    runtime: Arc::new(Workspace::new(
                        &root,
                        self.workspace_scratch(&root),
                        timeout,
                    )),
                },
            );
            state.last_activity = now;
            let runtime = Arc::clone(
                &state
                    .workspaces
                    .get(&root)
                    .expect("inserted workspace must exist")
                    .runtime,
            );
            return Ok(WorkspaceLease {
                service: self,
                root,
                runtime,
            });
        }
    }

    fn release(&self, root: &Path) {
        let mut state = self.lock_state();
        let now = Instant::now();
        if let Some(workspace) = state.workspaces.get_mut(root) {
            workspace.active = workspace.active.saturating_sub(1);
            workspace.last_used = now;
        }
        state.last_activity = now;
        self.workspace_available.notify_all();
    }

    fn lock_state(&self) -> MutexGuard<'_, DaemonState> {
        self.state
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner())
    }

    fn workspace_scratch(&self, root: &Path) -> PathBuf {
        let mut hasher = DefaultHasher::new();
        root.hash(&mut hasher);
        self.scratch_root.join(format!("{:016x}", hasher.finish()))
    }
}

impl Drop for WorkspaceLease<'_> {
    fn drop(&mut self) {
        self.service.release(&self.root);
    }
}

fn query_timeout(seconds: f64) -> Duration {
    let seconds = if seconds.is_nan() {
        1.0
    } else {
        seconds.clamp(1.0, 3600.0)
    };
    Duration::from_secs_f64(seconds)
}

pub fn execute(cli: &Cli, root: &Path, transport: &'static str) -> Value {
    let started = Instant::now();
    if let Some(failure) = boundary::evaluate(
        cli,
        root,
        started.elapsed().as_secs_f64() * 1000.0,
        transport,
    ) {
        return serde_json::to_value(failure)
            .expect("boundary response serialization must succeed");
    }

    if let Command::Find(arguments) = &cli.command
        && (arguments.text || arguments.mode == FindMode::Text)
    {
        return execute_text_search(cli, arguments, root, transport, started);
    }
    let scratch = std::env::temp_dir().join(format!("codeq-2-oneshot-{}", std::process::id()));
    let workspace = Workspace::new(root, scratch.clone(), query_timeout(cli.timeout));
    let result = execute_with_workspace(cli, root, transport, &workspace);
    workspace.close();
    let _ = fs::remove_dir_all(scratch);
    result
}

fn execute_with_workspace(
    cli: &Cli,
    root: &Path,
    transport: &'static str,
    workspace: &Workspace,
) -> Value {
    let started = Instant::now();
    if let Some(failure) = boundary::evaluate(cli, root, 0.0, transport) {
        return serde_json::to_value(failure)
            .expect("boundary response serialization must succeed");
    }
    if let Command::Find(arguments) = &cli.command
        && (arguments.text || arguments.mode == FindMode::Text)
    {
        return execute_text_search(cli, arguments, root, transport, started);
    }

    let sessions_before = workspace.session_stats();
    let metrics_before = workspace.metrics();
    let execution_started = Instant::now();
    let produced = match &cli.command {
        Command::Context(arguments) => semantic::context(workspace, arguments, cli.limit),
        Command::Trace(arguments) => semantic::trace(workspace, arguments, cli.limit),
        Command::Find(arguments)
            if arguments.mode == FindMode::Concept
                || (arguments.mode == FindMode::Auto
                    && concept::has_multiple_terms(&arguments.query)) =>
        {
            concept::search(workspace, arguments, cli.limit)
        }
        Command::Find(arguments) => semantic::find(workspace, arguments, cli.limit),
        Command::Review(arguments) => gitreview::review(workspace, arguments, cli.limit),
    };
    let mut data = produced.unwrap_or_else(|error| {
        serde_json::json!({
            "status": "error",
            "target": cli.command.target(),
            "error": error,
        })
    });
    let phase_ms = data
        .as_object_mut()
        .and_then(|object| object.remove("_phase_ms"));
    let sessions = workspace.session_stats();
    let metrics = workspace.metrics();
    let mut meta = serde_json::json!({
        "root": workspace.root(),
        "duration_ms": started.elapsed().as_secs_f64() * 1000.0,
        "queue_ms": 0.0,
        "execution_ms": execution_started.elapsed().as_secs_f64() * 1000.0,
        "lsp_sessions_before": sessions_before,
        "lsp_sessions": sessions,
        "lsp_started": metrics.sessions_started > metrics_before.sessions_started,
        "lsp_request_count": metrics.lsp_request_count.saturating_sub(metrics_before.lsp_request_count),
        "prewarm_files": metrics.prewarm_files.saturating_sub(metrics_before.prewarm_files),
        "prewarm_probes": metrics.prewarm_probes.saturating_sub(metrics_before.prewarm_probes),
        "prewarm_early_stops": metrics.prewarm_early_stops.saturating_sub(metrics_before.prewarm_early_stops),
        "cache": {
            "document_symbols_hit": metrics.document_symbols_hit.saturating_sub(metrics_before.document_symbols_hit),
            "document_symbols_miss": metrics.document_symbols_miss.saturating_sub(metrics_before.document_symbols_miss),
            "document_symbols_waited": metrics.document_symbols_waited.saturating_sub(metrics_before.document_symbols_waited),
            "document_symbols_evicted": metrics.document_symbols_evicted.saturating_sub(metrics_before.document_symbols_evicted),
            "document_symbol_entries": metrics.document_symbol_entries,
        },
        "transport": transport,
    });
    if let Some(phase_ms) = phase_ms {
        meta["phase_ms"] = phase_ms;
    }
    if let Some(lexical) = data.get("lexical_references") {
        meta["text"] = serde_json::json!({
            "matching_file_count": integer(lexical, "matching_file_count"),
            "tracked_matching_lines": integer(lexical, "tracked_line_count"),
            "untracked_matching_lines": integer(lexical, "untracked_line_count"),
        });
    }
    data["_meta"] = meta;
    data["schema_version"] = Value::from(SCHEMA_VERSION);
    data
}

fn execute_text_search(
    cli: &Cli,
    arguments: &crate::cli::FindArgs,
    root: &Path,
    transport: &'static str,
    started: Instant,
) -> Value {
    let mut data = match textsearch::search(
        root,
        &arguments.query,
        cli.limit,
        &arguments.paths,
        &arguments.globs,
        arguments.exclude_tests,
    ) {
        Ok(data) => data,
        Err(error) => serde_json::json!({
            "status": "error",
            "mode": "text",
            "search_mode": "text",
            "query": arguments.query,
            "reason": error,
            "results": [],
        }),
    };
    data["search_mode"] = Value::String("text".to_owned());
    let duration_ms = started.elapsed().as_secs_f64() * 1000.0;
    data["_meta"] = serde_json::json!({
        "root": root,
        "duration_ms": duration_ms,
        "queue_ms": 0.0,
        "execution_ms": duration_ms,
        "lsp_sessions_before": [],
        "lsp_sessions": [],
        "lsp_started": false,
        "lsp_request_count": 0,
        "prewarm_files": 0,
        "prewarm_probes": 0,
        "prewarm_early_stops": 0,
        "cache": {
            "document_symbols_hit": 0,
            "document_symbols_miss": 0,
            "document_symbols_waited": 0,
            "document_symbols_evicted": 0,
            "document_symbol_entries": 0,
        },
        "text": {
            "matching_file_count": data.get("matching_file_count").and_then(Value::as_u64).unwrap_or(0),
            "tracked_matching_lines": data.get("tracked_line_count").and_then(Value::as_u64).unwrap_or(0),
            "untracked_matching_lines": data.get("untracked_line_count").and_then(Value::as_u64).unwrap_or(0),
        },
        "transport": transport,
    });
    data["schema_version"] = Value::from(SCHEMA_VERSION);
    data
}

fn integer(data: &Value, key: &str) -> u64 {
    data.get(key).and_then(Value::as_u64).unwrap_or(0)
}
